tools/cloud_storage_viewer/src/deltafor/mod.rs

```diff
@@ -0,0 +1,251 @@
+        DeltaDelta {
+            step_size: step.unwrap_or(0),
+        }
+    }
+    fn encode(&mut self, last: i64, row: &[i64; 16], buf: &mut [i64; 16]) -> u8 {
+        let mut p = last;
+        let mut agg: i64 = 0;
+        for i in 0..16 {
+            assert!(
+                row[i] >= p,
+                "Value {} can't be larger than the previous one {}",
+                row[i],
+                p
+            );
+            let delta = row[i] - p;
+            assert!(
+                delta >= self.step_size,
+                "Delta {} can't be larger than step size {}",
+                delta,
+                self.step_size
+            );
+            buf[i] = (row[i] - p) - self.step_size;
+            agg |= buf[i];
+            p = row[i];
+        }
+        return 64 - (agg as u64).leading_zeros() as u8;
+    }
+    fn decode(&mut self, init: i64, row: &mut [i64; 16]) -> i64 {
+        let mut p = init;
+        for i in 0..16 {
+            row[i] = row[i] + p + self.step_size;
+            p = row[i];
+        }
+        return p;
+    }
+}
+
+#[allow(dead_code)]
+pub struct DeltaFOREncoder<Alg: DeltaAlg> {
+    delta_alg: Alg,
+    count: u64,
+    last: i64,
+}
+
+#[allow(dead_code)]
+impl<Alg: DeltaAlg> DeltaFOREncoder<Alg> {
+    pub fn new(init: i64, step: Option<i64>) -> Self {
+        DeltaFOREncoder {
+            delta_alg: Alg::new(step),
+            count: 0,
+            last: init,
+        }
+    }
+
+    pub fn add_row(&mut self, row: &[i64; 16], data: &mut dyn io::Write) -> Result<(), Error> {
+        let mut buf: [i64; 16] = [0; 16];
+        let nbits = self.delta_alg.encode(self.last, row, &mut buf);
+        self.last = row[15];
+        data.write_all(&nbits.to_le_bytes())?;
+        let mut tmp = buf.map(|x| x as u64);
+        bitpack::pack(&mut tmp, nbits as u32, data)?;
+        self.count += 1;
+        Ok(())
+    }
+}
+
+pub struct DeltaFORDecoder<Alg: DeltaAlg> {
+    delta_alg: Alg,
+    count: u64,
+    initial: i64,
+    pos: u64,
+}
+
+impl<Alg: DeltaAlg> DeltaFORDecoder<Alg> {
+    pub fn new(count: u64, initial_value: i64, step: Option<i64>) -> Self {
+        DeltaFORDecoder {
+            delta_alg: Alg::new(step),
+            count: count,
+            initial: initial_value,
+            pos: 0,
+        }
+    }
+
+    pub fn read_row(&mut self, row: &mut [i64; 16], data: &mut dyn io::Read) -> Result<(), Error> {
+        if self.pos == self.count {
+            return Err(Error::new(ErrorKind::UnexpectedEof, "EOF"));
+        }
+        let mut nbuf: [u8; 1] = [0; 1];
+        data.read_exact(&mut nbuf).unwrap();
+        let nbits = nbuf[0];
+        let mut buf: [u64; 16] = [0; 16];
+        bitpack::unpack(&mut buf, nbits as u32, data).unwrap();
+        *row = buf.map(|x| x as i64);
+        self.initial = self.delta_alg.decode(self.initial, row);
+        self.pos = self.pos + 1;
+        return Ok(());
+    }
+}
+
+pub fn read_index_header(
```

Comment:
Even though in this main() it is only use in synchronous code, it probably makes sense to write all our I/O functions as `async` with futures/tokio stream classes, so that they're usable in larger async programs later.

Reply:
What is the performance impact of `async` in case if the underlying `io::Read` is an in-memory buffer? The code uses `io::Read` a lot. Every bitpacking subroutine uses it. If every async call introduces a scheduling point it could be expensive. Maybe it makes sense to load the manifest or index into memory entirely and then use `io::Cursor` to pass the data to the decoder?

Reply:
The code actually uses `io::BufReader` so it's not in memory, but can be converted quickly depending on the impact of async.